### DISCUSSION TITLE
Alternative to Enums

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#521] [Q&A] [answered]
I'm creating a grammar with a concrete syntax that is html-like for attribute assignments, e.g.,:

`<Element attribute="value" />`

To be syntactically correct, the attribute's value should always be enclosed in quotes. There are cases where I want to limit _value_ to a range of pre-defined values. 

In Xtext, I'd have created an enum and assigned the enum's value to the matching string value, making the parser recognize the string value as the enum value. Since Langium requires me to create a data type rule for this, I'm not really sure how to accomplish it. I have defined the following data type rule as an example:

```
INVOCATION_CONDITION returns string:
    "ManualInvoke"      | "PostPropertyChange"  | 'PostRender'         |
    "PostUpdate"        | "PrePropertyChange"   | "PreRender"           |
    "PreUpdate"         | "CheckUpdate"         | "UserInteraction"     |
    "RenderDelegate"    | "UpdateDelegate"
    ;

```
and a rule that uses it like this:

```
Behavior:
    '<Behavior' 
        'id' '=' name=STRING
        'invocationCondition' '=' invocationCondition=INVOCATION_CONDITION
        'type' '='  type=STRING
        'package' '=' package=STRING(
    '/>' |
    '>'
        ('<Parameters>' parameters+=Parameter* '</Parameters>')?
    '</Behavior>')
    ;
```

But I get a parser error if I enclose a value for `INVOCATION_CONDITION` in quotes in the concrete syntax. It works if it's not enclosed in quotes. Is there a recommended way to accomplish what  I want to do here?

Thanks

**Comment by spoenemann** [comment, marked as answer]
In this case you could write it like this:
```
'invocationCondition' '=' '"' invocationCondition=INVOCATION_CONDITION '"'
```
But I'm not sure how this affects the lexer (tokenizer). Chevrotain has a simple lexing approach that takes the first match in the defined order, so it might be that this shadows the general `STRING` terminal rule. An alternative is to hard-code the quotes into the enum values:
```
'"ManualInvoke"'      | '"PostPropertyChange"'  | '"PostRender"'         | ...
```

**Reply by spoenemann** [reply]
Consequence of the second approach above: the string values in the AST include the quote characters as well. You can solve that with a custom ValueConverter.

Another option is to use the general `STRING` terminal instead of a data type rule, and restrict the possible values with a validation check.

**Reply by msujew** [reply]
I believe that the first approach would break the lexing completely, as strings won't be able to be parsed at all. The lexer will always prioritize the `"` token before any string. The proposed alternative would work. There's also another approach, where one overrides the `TokenBuilder` and wraps the keywords in `"` characters (the `LONGER_ALT` would need to be set accordingly).

**Reply by ballcoach12** [reply]
Thank you, gentlemen. This is a very helpful discussion that gives me several options.

For the short term, I have opted to go with specifying the values with the STRING terminal rule and enforcing the value through validation.
